ADR/0039-creating-pending-pipelineruns.md

```diff
@@ -0,0 +1,66 @@
+
+If we implement MPC limiting, it will make sure pipeline runs are not timeouting because of MPC pool being used by builds started earlier.
+
+### Example
+
+This is an example of controller that +- implements this MPC limiting log:
+
+```
+2024-11-20 14:04:44,646 prs_starter Thread-3 (my_capacity_counter) INFO Loaded MPC capacity: {'linux/arm64': 7, 'linux/ppc64le': 32, 'linux/s390x': 0}
+2024-11-20 14:04:49,659 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Reconciling my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90)
+2024-11-20 14:04:49,659 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Missing capacity for my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90) on linux/s390x
+2024-11-20 14:04:49,659 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Reconciling my-comp-3-on-push-hj5bw/jhutar-tenant (ba328d32-e912-4a2c-98e8-506937577912)
+2024-11-20 14:04:49,659 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Missing capacity for my-comp-3-on-push-hj5bw/jhutar-tenant (ba328d32-e912-4a2c-98e8-506937577912) on linux/s390x
+2024-11-20 14:04:57,309 prs_starter Thread-3 (my_capacity_counter) INFO Loaded MPC capacity: {'linux/arm64': 7, 'linux/ppc64le': 32, 'linux/s390x': 1}
+2024-11-20 14:04:59,659 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Reconciling my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90)
+2024-11-20 14:04:59,660 prs_starter.my_reconciler Thread-2 (my_reconciler) INFO Capacity available for my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90) on all considered platforms
+2024-11-20 14:04:59,660 prs_starter.my_reconciler Thread-2 (my_reconciler) INFO Starting my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90)
+2024-11-20 14:05:00,529 prs_starter.my_reconciler Thread-2 (my_reconciler) INFO The my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90) was started, not watching it anymore
+2024-11-20 14:05:00,529 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Reconciling my-comp-3-on-push-hj5bw/jhutar-tenant (ba328d32-e912-4a2c-98e8-506937577912)
+2024-11-20 14:05:00,529 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Missing capacity for my-comp-3-on-push-hj5bw/jhutar-tenant (ba328d32-e912-4a2c-98e8-506937577912) on linux/s390x
+```
+
+This is what was happening:
+
+* at 2024-11-20 14:04:44,646, we checked MPC capacity and s390x was fully used, so reconciling pipeline runs (that are in pending state) `my-comp-2-on-push-...` and `my-comp-3-on-push-...` did not started them
+* at 2024-11-20 14:04:57,309, we noticed MPC freed the capacity for s390x
+* at 2024-11-20 14:04:59,660 pipeline run `my-comp-2-on-push-...` was started, no luck for `my-comp-3-on-push-...` though
```

Comment:
These types of capacity decisions are not always blocked on single events/quotas. When considering the multi-platform controller, there are multiple different platforms that may be requested. 

* How can we reconcile the queues for all of these platforms?
* It seems like we would need to add some sort of queuing mechanism for the pending PipelineRuns. How can we ensure fairness with all of the pipelines which may be waiting for different metrics/resources to indicate capacity?
* How can we ensure that a pipeline is not unnecessarily blocked by an unneeded resource just because it is waiting behind some other pipeline in a queue?

Reply:
> How can we reconcile the queues for all of these platforms?

Here I was checking all platforms from that `prsstarter.example.com/mpc_platforms: linux/arm64,linux/ppc64le,linux/s390x` annotation, but `linux/s390x` was the only one that was not available. Code looks like this:

```
def can_pr_be_started(pr_required_platforms, mpc_available_slots)
    for platform in pr_required_platforms:
        if mpc_available_slots[platform] <= 0:
            return False
    return True
```

> It seems like we would need to add some sort of queuing mechanism for the pending PipelineRuns. How can we ensure fairness with all of the pipelines which may be waiting for different metrics/resources to indicate capacity?

You are right, we need to be reconciling (i.e. possibly starting) pipelineruns in order how they were created.

We can also introduce come sort of priority annotation that would also allow the pipelineruns to mess with the order (e.g. you are building security update and you do not want to wait for all ongoing rebuild of all RHEL 9.8 packages build).

> How can we ensure that a pipeline is not unnecessarily blocked by an unneeded resource just because it is waiting behind some other pipeline in a queue?

I was thinking about it in a way that each pipeline gated only by gating plugins it configures in it's annotations, so if I want to be gated by MPC, I use `prsstarter.example.com/mpc_platforms`, if I want to be gated by something else, I use `prsstarter.example.com/some_other_plugin`

And we can also have some global, implicit gateways like that etcd protection one that would be enabled for all pending pipelines by default.

Of course this all would only apply to pipeline runs created in this Pending state, so normal pipelines would just go ahead with kinda priority, as they would not be restricted by this gating.

Reply:
>  It seems like we would need to add some sort of queuing mechanism for the pending PipelineRuns. How can we ensure fairness with all of the pipelines which may be waiting for different metrics/resources to indicate capacity?

We could do a simple "FIFO" based on the creation dates.

Reply:
This is certainly an option, but it would result in a single-arch build staying in pending until multi-arch builds at the queue HEAD have sufficient capacity scheduled? 

Or would there be separate queues for drastically different types of requests?